index.ts

```diff
@@ -141,97 +146,124 @@ function isDelegator(userId: string, event: MetadataEvent): boolean {
     
-            // TODO Send delete nostr event to rely(s)
+    const headers = new Headers();
+    object.writeHttpMetadata(headers);
+    headers.set("etag", object.httpEtag);
+    return new Response(object.body, { headers });
+  },
+  'PUT@/file/:hash': restricted(["admin", "user"], async function putFile(request, { env, authEvent, workerContext }, params) {
+    if(!params.hash.match(hexKeyPattern)) return new Response(`Invalid SHA256 hash: ${params.hash}`, { status: 400 });
+    if(!request.body) return new Response("Missing body", { status: 400 });
+
+    // We expect the SHA256 hash from the URL to match the one of the request body (ie. the uploaded file).
+    // Per NIP-98 the authentication event SHOULD include a SHA256 hash of the request body in a 'payload' tag as hex (["payload", "<sha256-hex>"]).
+    // We will validate that the actual payload hash match when uploading to R2.
+    const payloadTag = authEvent.tags.find((t: string[]) => t[0] === 'payload')
+    if(payloadTag && payloadTag[1] && payloadTag[1] != params.hash) {
+      return new Response(`Unauthorized: Invalid nostr event, payload signature does not match URL`, { status: 401 });
+    }
 
-            return new Response(null, { status: 204 }) // Success response with no content.
-          
+    try {
+      if(await env.BANBOORU_BUCKET.head(params.hash)) {
+        return new Response("Conflict", { status: 409 })
+      }
+      const object = await env.BANBOORU_BUCKET.put(params.hash, request.body, {
+        // See https://developers.cloudflare.com/r2/api/workers/workers-api-reference/#r2putoptions
+        // R2 will check the received data SHA-256 hash to confirm received object integrity and will fail if it does not.
+        sha256: params.hash,
+        httpMetadata: {
+          contentType: request.headers.get('content-type') || undefined,
+          cacheControl: 'public, max-age=31536000, immutable' // let's client cache this for a year
+        }
+      });
+      if (!object) {
+        if(payloadTag && payloadTag[1]) { // The user signed the hash as part of the auth event. 
+          return new Response(`Unauthorized: Invalid nostr event, payload signature invalid`, { status: 401 });
         } else {
-          return new Response("Forbidden.", { status: 403 });
+          return new Response("Bad Request", { status: 400 })
         }
-      } catch (error) {
-        console.error("Object DEL failed", error);
-        return new Response('Server error', { status: 500 });
-      } 
-    })
+      }
+
+      // Handle metadata without blocking the response.
+      workerContext.waitUntil(handleMetadada(
+        `https://${request.headers.get('host')}`,
+          object,
+          env,
+          // TODO Safer parsing of the NIP Delegation
+          request.headers.has("x-nip-26-delegation") ? JSON.parse(request.headers.get("x-nip-26-delegation") as string) as nip26.Delegation : undefined
```

Comment:
is this supposed to be `|` instead of `:` at the end of the line?
```
request.headers.has("x-nip-26-delegation") ? JSON.parse(request.headers.get("x-nip-26-delegation") as string) as nip26.Delegation | undefined
```

Reply:
The coe is the same as
```
if (request.headers.has("x-nip-26-delegation")) {
  JSON.parse(request.headers.get("x-nip-26-delegation") as string) as nip26.Delegation
} else {
  undefined
}
```
But on a single line. It's a but of an abuse of the ternary operator but this part needs more love anyway